### PR TITLE
Refactor the scheduler interface

### LINKED ISSUE #1674: Refactor the scheduler interface
Currently, we have a namespace `caf::scheduler`. However, a single interface `caf::scheduler` would suffice. The two implementations (work stealing and work sharing) should be implementation details, not implemented in the public API.

### PR DESCRIPTION
**Pull request by riemass**
Closes #1674.

Hiding the `coordinator` in cpp pulled a couple of other classes into the cpp, since they are in a `circular` template dependency (`circular` for the lack of a better name). I've also cleaned up headers on all the files I've touched and removed a lot of legacy files. 
@Neverlord  What do you think if we should loose the `policy` template specialization and implement `work_sharing` and `work_stealing` as complete classes?

### CODE REVIEW

File: libcaf_core/caf/local_actor.hpp
Comment (Neverlord) [inline libcaf_core/caf/local_actor.hpp:1]: None of these changes have anything to do with the PR. I suspect these are tool-found? Mind filing these findings separately? Same for the other files that were only touched to re-organize includes.

File: libcaf_core/caf/scheduler.cpp [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+
+#include <condition_variable>
+#include <fstream>
+#include <ios>
+#include <iostream>
+#include <memory>
+#include <random>
+#include <thread>
+
+#define CONFIG(str_name, var_name)                                             \
```
Comment (Neverlord) [inline libcaf_core/caf/scheduler.cpp:29]: There are multiple implementations, but this seems to be for the work-stealing implementation specifically. In any case, all calls of this macro are in a single function, so a lambda would be better.

File: libcaf_core/caf/scheduler.cpp [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+  size_t id_;
+  // pointer to central coordinator
+  scheduler_impl_ptr parent_;
+  // policy-specific data
+  policy_data data_;
+  // instance of our policy object
+  Policy policy_;
+};
+
+/// Policy-based implementation of the abstract coordinator base class.
```
Comment (Neverlord) [inline libcaf_core/caf/scheduler.cpp:310]: There no longer is an abstract coordinator.

File: libcaf_core/caf/scheduler.hpp [hunk 3]
```diff
@@ -4,6 +4,85 @@
...
+/// A coordinator creates the workers, manages delayed sends and
+/// the central printer instance for {@link aout}. It also forwards
+/// sends from detached workers or non-actor threads to randomly
+/// chosen workers.
+class CAF_CORE_EXPORT scheduler : public actor_system::module {
+public:
+  // -- factory functions ------------------------------------------------------
+  static scheduler* make_work_stealing(actor_system& sys);
+
+  static scheduler* make_work_sharing(actor_system& sys);
```
Comment (Neverlord) [inline libcaf_core/caf/scheduler.hpp:26]: Should not return raw pointers (or not be public).

File: libcaf_core/caf/scheduler.hpp [hunk 4]
```diff
@@ -4,6 +4,85 @@
...
+
+  static void cleanup_and_release(resumable*);
+
+  virtual detail::actor_local_printer_ptr printer_for(local_actor* self);
+
+  // -- properties -------------------------------------------------------------
+
+  static size_t default_thread_count() noexcept {
+    return std::max(std::thread::hardware_concurrency(), 4u);
+  }
```
Comment (Neverlord) [inline libcaf_core/caf/scheduler.hpp:55]: I think this should move to the `.cpp` file, even if it's a small function. If we ever change the default around, we want linked applications to use the new default.

File: libcaf_core/tests/legacy/thread_hook.cpp [hunk 5]
```diff
@@ -99,22 +101,24 @@ BEGIN_FIXTURE_SCOPE(fixture<counting_thread_hook>)
...
-  auto fallback = scheduler::abstract_coordinator::default_thread_count();
+  auto fallback = scheduler::default_thread_count();
   assumed_thread_count = get_or(cfg, "caf.scheduler.max-threads", fallback)
                          + 2; // clock and private thread pool
-  auto& sched = sys.scheduler();
-  if (sched.detaches_utility_actors())
-    assumed_thread_count += sched.num_utility_actors();
+                              // auto& sched = sys.scheduler();
+                              // if (sched.detaches_utility_actors())
+  //   assumed_thread_count += sched.num_utility_actors();
```
Comment (Neverlord) [inline libcaf_core/tests/legacy/thread_hook.cpp:109]: Did you mean to remove this?

File: libcaf_core/tests/legacy/thread_hook.cpp [hunk 6]
```diff
@@ -99,22 +101,24 @@ BEGIN_FIXTURE_SCOPE(fixture<counting_thread_hook>)
...
-  auto fallback = scheduler::abstract_coordinator::default_thread_count();
+  auto fallback = scheduler::default_thread_count();
   assumed_thread_count = get_or(cfg, "caf.scheduler.max-threads", fallback)
                          + 3; // clock, private thread pool, and  detached actor
-  auto& sched = sys.scheduler();
-  if (sched.detaches_utility_actors())
-    assumed_thread_count += sched.num_utility_actors();
+                              // auto& sched = sys.scheduler();
+                              // if (sched.detaches_utility_actors())
+  // assumed_thread_count += sched.num_utility_actors();
```
Comment (Neverlord) [inline libcaf_core/tests/legacy/thread_hook.cpp:120]: Did you mean to remove this?

File: libcaf_io/caf/io/basp/worker.hpp [hunk 7]
```diff
@@ -35,7 +33,7 @@ class CAF_IO_EXPORT worker : public detail::abstract_worker,
 
   using super = detail::abstract_worker;
 
-  using scheduler_type = scheduler::abstract_coordinator;
+  using scheduler_type = scheduler;
```
Comment (Neverlord) [inline libcaf_io/caf/io/basp/worker.hpp:36]: The alias is longer than the actual type. 😉

I'd drop it.

**Review by Neverlord — APPROVED** [review]
Just some nitpicks.

File: libcaf_core/caf/scheduler.cpp (hunk 1 above)
Comment (riemass) [reply libcaf_core/caf/scheduler.cpp:29]: Calls are in the initializer list so we cant make a lambda here, but I'll inline the calls, doesn't look bad.

File: libcaf_core/caf/scheduler.hpp (hunk 4 above)
Comment (riemass) [reply libcaf_core/caf/scheduler.hpp:55]: This was in a cpp but I inlined it thinking it's a short function. Moving back.

File: libcaf_core/tests/legacy/thread_hook.cpp (hunk 5 above)
Comment (riemass) [reply libcaf_core/tests/legacy/thread_hook.cpp:109]: Yes, tested and forgot to remove.

File: libcaf_core/caf/local_actor.hpp
Comment (riemass) [reply libcaf_core/caf/local_actor.hpp:1]: Cleaned up 11 files, will follow up.

File: libcaf_io/caf/io/basp/worker.hpp (hunk 7 above)
Comment (riemass) [reply libcaf_io/caf/io/basp/worker.hpp]: Seems it wasn't used.

**Comment by Neverlord** [comment]
> @Neverlord What do you think if we should loose the policy template specialization and implement work_sharing and work_stealing as complete classes?

@riemass all for it, but let's do that as a followup.

**Review by Neverlord — APPROVED** [review]

**Pull request merged**